components/layout_2020/flow/root.rs

```diff
@@ -229,15 +200,165 @@ impl FragmentTreeRoot {
+        iterator.border_dimensions
+    }
+}
+
+/// A top-down fragment border box iteration handler.
+trait FragmentIterator {
+    /// The operation to perform.
+    fn process(&mut self, fragment: &Fragment, containing_block: &PhysicalRect<Length>) -> bool;
+}
+
+struct FragmentContentBoxIterator {
+    node_address: OpaqueNode,
+    rect: PhysicalRect<Length>,
+}
+
+impl FragmentContentBoxIterator {
+    fn new(node_address: OpaqueNode) -> FragmentContentBoxIterator {
+        FragmentContentBoxIterator {
+            node_address,
+            rect: PhysicalRect::zero(),
+        }
+    }
+}
+
+impl FragmentIterator for FragmentContentBoxIterator {
+    fn process(&mut self, fragment: &Fragment, containing_block: &PhysicalRect<Length>) -> bool {
+        let fragment_relative_rect = match fragment {
+            Fragment::Box(fragment) if fragment.tag == self.node_address => fragment
+                .border_rect()
+                .to_physical(fragment.style.writing_mode, &containing_block),
+            Fragment::Text(fragment) if fragment.tag == self.node_address => fragment
+                .rect
+                .to_physical(fragment.parent_style.writing_mode, &containing_block),
+            Fragment::Box(_) | Fragment::Text(_) | Fragment::Image(_) | Fragment::Anonymous(_) => {
+                return true
+            },
+        };
+
+        self.rect = fragment_relative_rect.translate(containing_block.origin.to_vector());
+        return false;
+    }
+}
+
+struct FragmentBorderDimensionsIterator {
+    node_address: OpaqueNode,
+    border_dimensions: Rect<i32>,
+}
+
+impl FragmentBorderDimensionsIterator {
+    fn new(node_address: OpaqueNode) -> FragmentBorderDimensionsIterator {
+        FragmentBorderDimensionsIterator {
+            node_address,
+            border_dimensions: Rect::zero(),
+        }
+    }
+}
+
+impl FragmentIterator for FragmentBorderDimensionsIterator {
+    fn process(&mut self, fragment: &Fragment, containing_block: &PhysicalRect<Length>) -> bool {
+        let (style, border_box) = match fragment {
+            Fragment::Box(fragment) if fragment.tag == self.node_address => {
+                (&fragment.style, fragment.border_rect())
+            },
+            Fragment::Box(_) | Fragment::Text(_) | Fragment::Image(_) | Fragment::Anonymous(_) => {
+                return true
+            },
+        };
+
+        // https://drafts.csswg.org/cssom-view/#dom-element-clienttop
+        // " If the element has no associated CSS layout box or if the
+        //   CSS layout box is inline, return zero." For this check we
+        // also explicitly ignore the list item portion of the display
+        // style.
+        let display = &style.get_box().display;
+        if display.inside() == style::values::specified::box_::DisplayInside::Flow &&
+            display.outside() == style::values::specified::box_::DisplayOutside::Inline
+        {
+            return true;
```

Comment:
At this point we’ve found a fragment generated by the appropriate DOM element but it had `display: inline`. Why continue the tree iteration? Aren’t other fragments generated by that element also gonna be inline?

Reply:
I think you are right. I'll change this to a return `false` instead.